File: src/java/2_CreationalPatterns/1_Builder/BuilderExercise.java
package com.activemesa.creational.builder.exercise;

import org.junit.Test;
import org.junit.Assert;

import java.util.ArrayList;
import java.util.List;

class Field
{
  public String type, name;

  public Field(String name, String type) {
    this.type = type;
    this.name = name;
  }

  @Override
  public String toString() {
    return String.format("public %s %s;", type, name);
  }
}

class Class
{
  public String name;
  public List<Field> fields = new ArrayList<>();

  public Class(){}

  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    String nl = System.lineSeparator();
    sb.append("public class " + name).append(nl)
      .append("{").append(nl);
    for (Field f : fields)
      sb.append("  " +  f).append(nl);
    sb.append("}").append(nl);
    return sb.toString();
  }
}

class CodeBuilder
{
  private Class theClass = new Class();

  public CodeBuilder(String rootName)
  {
    theClass.name = rootName;
  }

  public CodeBuilder addField(String name, String type)
  {
    theClass.fields.add(new Field(name, type));
    return this;
  }

  @Override
  public String toString() {
    return theClass.toString();
  }
}

//import org.junit.Test;
//import org.junit.Assert;
//import com.udemy.ucp.*;

File: src/js/2_CreationalPatterns/1_Builder/builder_exercise.js
class Field
{
  constructor(name)
  {
    this.name = name;
  }
}

class Class
{
  constructor(name)
  {
    this.name = name;
    this.fields = [];
  }

  toString()
  {
    let buffer = [];
    buffer.push(`class ${this.name} {\n`);

    if (this.fields.length > 0) {
      buffer.push(`  constructor(`);
      for (let i = 0; i < this.fields.length; ++i) {
        buffer.push(this.fields[i].name);
        if (i + 1 !== this.fields.length)
          buffer.push(', ');
      }
      buffer.push(`) {\n`);
      for (let field of this.fields) {
        buffer.push(`    this.${field.name} = ${field.name};\n`);
      }
      buffer.push('  }\n');
    }

    buffer.push('}');
    return buffer.join('');
  }
}

class CodeBuilder
{
  constructor(className)
  {
    this._class = new Class(className);
  }

  addField(name)
  {
    this._class.fields.push(
      new Field(name)
    );
    return this;
  }

  toString()
  {
    return this._class.toString();
  }
}

describe('builder', function()
{
  it('empty test', function()
  {
    let cb = new CodeBuilder('Foo');
    expect(cb.toString())
      .toEqual('class Foo {\n}');
  });

  it('person test', function()
  {
    let cb = new CodeBuilder('Person');
    cb.addField('name').addField('age');
    expect(cb.toString())
      .toEqual('class Person {\n'
        + '  constructor(name, age) {\n'
        + '    this.name = name;\n'
        + '    this.age = age;\n'
        + '  }\n'
        +'}');
  });
});